File: Stack/RedundantBraces.cpp
/*
Given a string A denoting an expression. It contains the following operators ’+’, ‘-‘, ‘*’, ‘/’.

Chech whether A has redundant braces or not.

Return 1 if A has redundant braces, else return 0.

Note: A will be always a valid expression.
*/

#include <iostream>
#include <stack>
using namespace std;

int main(){
    string str;
    cin>> str;
    stack<char> st;
    bool result = false;
    for(int i=0; i<str.size(); i++){
        if(str[i] == '+' || str[i] == '-' || str[i] == '*' || str[i] == '/'){
            st.push(str[i]);
        }
        else if(str[i] == '('){
            st.push(str[i]);
        }
        else if(str[i] == ')'){
            if(st.top() == '('){
                result = true;
            }
            while(!st.empty() && st.top() != '('){
                st.pop();
            }
            st.pop();
        }
    }
    cout<<result;
    return 0;
}

File: Stack/RedundantBraces.java

import java.util.Scanner;
import java.util.Stack;


public class RedundantBraces {
 public static int braces(String a) {
    Stack<Character> st = new Stack<Character>();
    for(int i=0; i<a.length(); i++){
        if(a.charAt(i) == '+' || a.charAt(i) == '-' || a.charAt(i) == '*' || a.charAt(i) == '/'){
            st.push(a.charAt(i));
        }
        else if(a.charAt(i) == '('){
            st.push(a.charAt(i));
        }
        else if(a.charAt(i) == ')'){
            if(st.peek() == '('){
                return 1;
            }
            while(!st.empty() && st.peek()!='('){
            st.pop();
            }
            st.pop();   
        }
    }
    return 0;
}    
 public static void main(String args[]){
     Scanner scan = new Scanner(System.in);
     String a = scan.next();
     System.out.print(braces(a));
 }
}
